nixos/synapse: add support for workers, cleanup

999eagle opened the pull request:
###### Description of changes

This PR mainly adds support for running workers to matrix-synapse. Additionally, listener options are updated, an option for automatically configuring redis is added, and replication settings between workers and the main process are configured automatically. Workers are also being run as separate services with a common `matrix-synapse.target` to start the main process and all workers.

This is also running on my own matrix server as of https://git.catgirl.cloud/999eagle/dotfiles-nix/-/commit/59e1e0d79d03b603a09bc586a89704171e7463f3.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [X] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

999eagle committed:
nixos/synapse: update listener settings

The resource type health is currently missing, but should be available
according to https://matrix-org.github.io/synapse/latest/usage/configuration/config_documentation.html#listeners

999eagle committed:
nixos/synapse: cleanup, split out listener type and service config

999eagle committed:
nixos/synapse: add options to configure workers

999eagle committed:
nixos/synapse: add option to configure redis automatically

999eagle committed:
nixos/synapse: automatically configure replication listener

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -719,66 +814,128 @@ in {
+              '';
+              serviceConfig = {
+                ExecStartPre = [
+                  ("+" + (pkgs.writeShellScript "matrix-synapse-fix-permissions" ''
+                    chown matrix-synapse:matrix-synapse ${cfg.settings.signing_key_path}
+                    chmod 0600 ${cfg.settings.signing_key_path}
+                  ''))
```
Are we reducing the permissions? if so that is not super secure.

999eagle replied:
I haven't changed this part of the module, but I don't really see why we shouldn't chmod it to 0600?

mweinelt replied:
This chmod exists because once upon a time the key files were created with too wide permissions.

79e675444caf7b491b2c0d25277b046d3f6d8e04

SuperSandro2000 replied:
We could deprecate this based on stateVersion but that's something for another PR.

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -719,66 +814,128 @@ in {
+
+    services.redis.servers.matrix-synapse = lib.mkIf cfg.configureRedisLocally {
+      enable = true;
+      port = 0;
```
```suggestion
```
that's the default

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -719,66 +814,128 @@ in {
+            ];
+          };
+      in
+      lib.mkMerge ([
+        {
+          matrix-synapse = lib.mkMerge [
+            baseServiceConfig
```
I'm not sure whether that'd be faster but personally I feel like the current way looks a lot cleaner. With only one `mkMerge` it would look something like this:

```nix
mkMerge ([
  { matrix-synapse = baseServiceConfig; }
  {
    matrix-synapse = {
      ...
    };
  }
]
++ (builtins.map ...));
```

999eagle committed:
nixos/synapse: add test for running synapse with workers

Co-authored-by: Daniel Olsen <daniel.olsen99@gmail.com>

999eagle committed:
nixos/synapse: document options better

Co-authored-by: Daniel Olsen <daniel.olsen99@gmail.com>

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -666,13 +680,119 @@ in {
+                  worker_name = lib.mkOption {
+                    type = types.str;
+                    description = "Unique name for this worker";
+                  };
```
Done

999eagle committed:
nixos/synapse: remove obsolete log context

see https://github.com/matrix-org/synapse/commit/0304ad0c3d79e44e78f9658e71f1e1533e3aa4e2 for when this was removed upstream

999eagle committed:
nixos/synapse: automatically configure logging for synapse and workers

999eagle commented:
Latest changes are deployed on my Matrix server with https://git.catgirl.cloud/999eagle/dotfiles-nix/-/commit/bd63d09c73a81580bf2c99033018e0175fc05657.

999eagle commented:
Rebased onto `nixpkgs-unstable` to resolve conflicts introduced through #221318. Changes are also deployed on my server with https://git.catgirl.cloud/999eagle/dotfiles-nix/-/commit/7672239df3994a2bf69bcbf03fb469093759d700.

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -727,13 +784,124 @@ in {
+                  listener yourself.
+
+                  A redis server is required for running workers. A local one can be enabled
+                  using `services.matrix-synapse.configureRedisLocally`.
```
I think it might also be reasonable to explain that this is the place where to add worker-specific config that overrides stuff from the "shared" config (i.e. the declarations from `services.matrix-synapse.settings`). E.g. the shared config sets `enable_media_repo = false` whereas the media_repo worker's config sets it to `true` in here. This may be rather obvious when configuring synapse by hand because you specify multiple `-c` options with shared config and worker-specific config. This is however not transparent by this module (which is absolutely OK), but may cause to confusion (exhibit A: myself who was wondering how to do worker-specific config in here ;-) ).

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -784,64 +1000,123 @@ in {
+              ExecStart = ''
+                ${cfg.package}/bin/synapse_homeserver \
+                  ${ concatMapStringsSep "\n  " (x: "--config-path ${x} \\") ([ configFile ] ++ cfg.extraConfigFiles) }
+                  --keys-directory ${cfg.dataDir}
```
I haven't found a way to distribute workers over >1 machine. Any chance you know if that's reasonably possible? The main issue I see with that is that workers require the keys directory from synapse itself (and me being unwilling to use e.g. NFS).

If not, then we can resolve this right away. Otherwise I'd be curious on how to do that (given that I've mentioned this and possible problems with it in a few comments above) and what/if adjustments are needed.

999eagle replied:
See above, it's not really documented anywhere upstream how distributing workers over multiple machines might work but I'm fairly certain it should technically work. This might be a thing to add in another PR though as it requires quite a bit more work.

Ma27 replied:
Just discussed it directly, but for the record: none of us found documentation for that, so e.g. official support is entirely unclear. Because of that we consider it explicitly out of scope. Also it'd require even more changes to the module (i.e. to allow running workers without a main synapse).

Ma27 reviewed (commented):
I've successfully deployed this to my homeserver without any adjustments and it's working fine. There are a few rough edges though that I'd like to see cleaned up before merging. All in all thank you @999eagle and also @dali99 for your work on this (AFAIS both PRs seem to have a few inspirations from each other already).

Let me explain why I decided to go with this PR:
* I think the refactoring of the log config stuff to allow workers to configure logging on their own is quite nice
* There are rather helpful assertions in here which catch some easy-to-make mistakes.
* The entire generation of units for workers somehow works around the module-system in a very weird way that I really don't want to maintain.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -13,6 +13,7 @@ let
   usePostgresql = cfg.settings.database.name == "psycopg2";
   hasLocalPostgresDB = let args = cfg.settings.database.args; in
     usePostgresql && (!(args ? host) || (elem args.host [ "localhost" "127.0.0.1" "::1" ]));
+  hasWorkers = cfg.workers.enable && (cfg.workers.config != { });
```
The original idea was to have the usual "killswitch" most (all?) services have but in hindsight it's probably not very useful as a server requiring workers won't run with just the single main process.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -346,8 +490,8 @@ in {
               type = types.path;
-              default = ./synapse-log_config.yaml;
-              defaultText = lib.literalExpression "nixos/modules/services/matrix/synapse-log_config.yaml";
+              default = genLogConfigFile "synapse";
```
That's a good idea! We'd have to special-case this option during config file generation though to allow this as currently the config file is basically just `(pkgs.format.yaml {}).generate "homeserver.yaml" config.services.matrix-synapse.settings`. Do you think this would be worth it?

Ma27 replied:
Hah, totally misread that, sorry :sweat_smile: 

ftr my point still stands, I think the current changes to log_config are still an improvement.

My preferred solution would be something like `services.matrix-synapse.logConfig.defaults` (and then perhaps the possibility to define custom settings per worker-name, i.e. `services.matrix-synapse.logConfig.federation-sender`, just a first idea though). I think that's preferable over special-casing the format renderer because this seems like too much of a deviation from how this rfc42 stuff works in other modules which might be rather unexpected.

However, I'm resolving this now. As I said, I consider this state an improvement over the previous stuff already and more sophisticated changes to logging in a PR about workers would be a clear case of feature creep IMHO. Feel free to file a follow-up PR with that if you want to, I'll add it to my todolist, perhaps I'll get to it soonish otherwise :)

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -154,7 +197,108 @@ in {
+
+        tls = mkOption {
+          type = types.bool;
+          default = true;
```
There's no particular reason for this default as far as I can tell. I didn't change anything here, I just extracted the type definition for listeners to be able to reuse it for `services.matrix-synapse.settings.listeners` and `services.matrix-synapse.workers.*.worker_listeners`.

If workers are enabled, you have to terminate TLS at a reverse proxy anyway, so it'd probably make sense to default it to `false` anyway including for the main process. What do you think?

Ma27 replied:
> If workers are enabled, you have to terminate TLS at a reverse proxy anyway, so it'd probably make sense to default it to false anyway including for the main process. What do you think?

Correct. However, I think another common way to deploy is to actually expose port 8448 publicly (for federation in non-worker context) which needs to speak TLS then. Anyways, I don't really want to change the default for the main process's listener, I'm afraid that this might be too easy to miss on upgrade.

I agree however, there's no reason to do TLS for workers, so this default needs to change. I.e.

```nix
let
  listenerType = workerContext: types.submodule {
    options.tls = mkOption {
      default = !workerContext;
      /* ... */
    };
  };
in
```

and then `listenerType true` for `worker_listeners.type` and `listenerType false` for `listeners.type`.

999eagle replied:
Done.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -409,118 +553,31 @@ in {
-              } ];
+                }];
+              }] ++ lib.optional hasWorkers {
+                port = 9093;
```
There's no real suggested default AFAIK but the worker configuration docs use 9093 in their example: https://matrix-org.github.io/synapse/latest/workers.html#shared-configuration.

If we make it an option we wouldn't be able to move `services.matrix-synapse.workers.config` one level up like you suggested above. It might make sense to document the use of 9093 in the description of `services.matrix-synapse.workers` though.

I think it's technically possible to distribute worker processes over multiple machines but there is no documentation of any kind about this that I could find. Currently, this module doesn't really support worker distribution though because there's no (easy) way to run workers without running the main process as well. I think this could be done as another PR though.

Ma27 replied:
I'm fine with that considering that (1) I'd consider it the de-facto standard now considering the docs and (2) it's for internal communication only and we'd need to revise more stuff in this module anyways if we'd ever try to implement multi-machine worker setups (and I'm not sure I want that as long as there's no official documentation on it).

So, please mention port 9093 in e.g. the option's description and we're good.

999eagle replied:
Done.

999eagle committed:
nixos/synapse: move services.matrix-synapse.workers.config to services.matrix-synapse.workers

999eagle committed:
nixos/synapse: simplify replication listener assertion

999eagle commented on nixos/tests/matrix/synapse-workers.nix:
```diff
@@ -0,0 +1,53 @@
+    start_all()
+
+    homeserver.wait_for_unit("matrix-synapse.service");
+    homeserver.wait_for_unit("matrix-synapse-worker-federation_sender.service");
```
Looking through the upstream documentation I'm pretty sure we can't use the health check endpoint for this as it always returns 200 OK (yes, even if the redis connection fails) and I don't see any useful metrics that could be used for this either.

999eagle replied:
Done.

Ma27 replied:
> Done. 

I assume this was meant to be in another thread?

But yeah, seems as if there's not much to check unfortunately :/ 
I _may_ try out implementing a more detailed test, but this checks at least if everything starts up correctly and since there don't seem to be more meaningful healthchecks that are trivial to implement, let's keep it that way for now.

One thing though: can you please add this to `passthru.tests` of the synapse package? That way it gets built whenever the synapse package is touched.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -764,8 +941,47 @@ in {
+                listener:
+                  listener.port == main.port
+                  && (lib.any (resource: lib.any (name: name == "replication") resource.names) listener.resources)
+                  && (lib.any (bind: bind == main.host || bind == "0.0.0.0") listener.bind_addresses)
```
Added `::` here.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -764,8 +941,47 @@ in {
+              (
+                listener:
+                  listener.port == main.port
+                  && (lib.any (resource: lib.any (name: name == "replication") resource.names) listener.resources)
```
Changed

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -727,13 +784,124 @@ in {
+      workers = lib.mkOption {
+        default = { };
+        description = lib.mdDoc ''
+          Options for configuring workers. See `services.matrix-synapse.workers.enable`
```
I've changed that in the places I've found where it makes sense to do so.

999eagle committed:
nixos/synapse: default tls to off for workers and document worker replication port

999eagle committed:
matrix-synapse: add worker test to passthru.tests

Ma27 committed:
nixos/rl-2311: fix option references for synapse workers

Ma27 reviewed (approved):
LGTM, thanks!

I'll wait until tonight (in CEST) for comments from others, but then I'll merge since this has been open for two months now.

dali99 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -784,64 +988,123 @@ in {
+                description = "Synapse Matrix worker ${name}";
+                # make sure the main process starts first for potential database migrations
+                after = [ "matrix-synapse.service" ];
+                serviceConfig = {
```
https://github.com/NixOS/nixpkgs/pull/242026/files#diff-36428133f6120f9176e7210bcb4f75292a0c9cab54235840b13ba961b453d6d4R167-R171

This was never resolved and is relevant for this implementation as well

Ma27 replied:
Can you give me some instructions to reproduce the underlying issue? Because I cannot reproduce this at all.

Example 1 (`systemctl restart matrix-synapse.target`):

```
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix worker worker0...
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix worker worker1...
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix worker worker2...
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker1.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix worker worker1.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker1.service: Consumed 14min 23.993s CPU time, read 281.0K from disk, written 0B to disk, received 47.5M IP traffic, sent 101.7M IP traffic.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker2.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix worker worker2.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker2.service: Consumed 13min 42.896s CPU time, read 785.5K from disk, written 0B to disk, received 46.9M IP traffic, sent 94.6M IP traffic.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker0.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix worker worker0.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse-worker-worker0.service: Consumed 26min 15.411s CPU time, read 3.3M from disk, written 0B to disk, received 354.3M IP traffic, sent 165.7M IP traffic.
Sep 11 19:39:33 soost systemd[1]: Stopping Synapse Matrix homeserver...
Sep 11 19:39:33 soost systemd[1]: matrix-synapse.service: Deactivated successfully.
Sep 11 19:39:33 soost systemd[1]: Stopped Synapse Matrix homeserver.
Sep 11 19:39:33 soost systemd[1]: matrix-synapse.service: Consumed 45min 50.242s CPU time, read 78.9M from disk, written 0B to disk, received 153.6M IP traffic, sent 343.5M IP traffic.
Sep 11 19:39:33 soost systemd[1]: Starting Synapse Matrix homeserver...
Sep 11 19:39:44 soost systemd[1]: Started Synapse Matrix homeserver.
Sep 11 19:39:44 soost systemd[1]: Starting Synapse Matrix worker worker0...
Sep 11 19:39:44 soost systemd[1]: Starting Synapse Matrix worker worker1...
Sep 11 19:39:44 soost systemd[1]: Starting Synapse Matrix worker worker2...
Sep 11 19:39:49 soost systemd[1]: Started Synapse Matrix worker worker1.
Sep 11 19:39:50 soost systemd[1]: Started Synapse Matrix worker worker0.
Sep 11 19:39:50 soost systemd[1]: Started Synapse Matrix worker worker2.
```

Everything starts up properly most notably the workers start up in parallel (notice `Starting` vs `Started` _after_ synapse itself).

Same for the case where I restart all services individually:

```
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix homeserver...
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix worker worker0...
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix worker worker1...
Sep 11 19:42:10 soost systemd[1]: Stopping Synapse Matrix worker worker2...
Sep 11 19:42:10 soost systemd[1]: matrix-synapse-worker-worker0.service: Deactivated successfully.
Sep 11 19:42:10 soost systemd[1]: Stopped Synapse Matrix worker worker0.
Sep 11 19:42:10 soost systemd[1]: matrix-synapse-worker-worker0.service: Consumed 8.753s CPU time, received 573.3K IP traffic, sent 353.0K IP traffic.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker1.service: Deactivated successfully.
Sep 11 19:42:11 soost systemd[1]: Stopped Synapse Matrix worker worker1.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker1.service: Consumed 9.644s CPU time, received 328.4K IP traffic, sent 400.9K IP traffic.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker2.service: Deactivated successfully.
Sep 11 19:42:11 soost systemd[1]: Stopped Synapse Matrix worker worker2.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse-worker-worker2.service: Consumed 9.393s CPU time, received 308.8K IP traffic, sent 361.6K IP traffic.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse.service: Deactivated successfully.
Sep 11 19:42:11 soost systemd[1]: Stopped Synapse Matrix homeserver.
Sep 11 19:42:11 soost systemd[1]: matrix-synapse.service: Consumed 30.249s CPU time, received 985.8K IP traffic, sent 1.1M IP traffic.
Sep 11 19:42:11 soost systemd[1]: Starting Synapse Matrix homeserver...
Sep 11 19:42:19 soost systemd[1]: Started Synapse Matrix homeserver.
Sep 11 19:42:19 soost systemd[1]: Starting Synapse Matrix worker worker0...
Sep 11 19:42:19 soost systemd[1]: Starting Synapse Matrix worker worker1...
Sep 11 19:42:19 soost systemd[1]: Starting Synapse Matrix worker worker2...
Sep 11 19:42:25 soost systemd[1]: Started Synapse Matrix worker worker0.
Sep 11 19:42:26 soost systemd[1]: Started Synapse Matrix worker worker2.
Sep 11 19:42:26 soost systemd[1]: Started Synapse Matrix worker worker1.
```

dali99 replied:
I have conflicting information about this: On one hand this _should_ not be necessary, but in real deployments we have noticed that without it synapse workers will fail whenever there's a schema upgrade and you run `nixos-rebuild switch`

Solutions as I see it is either that script, or we could change some of the systemd startLimit stuff to either make it wait longer between restarts (so it doesn't reach it's burstLimit) or adjust the burst limits so that it just keeps trying.

I prefer the wait for script

Alternatively we could assume we're wrong in our observations and that this is a non-issue, but I suspect we're going to end up with adding something similar here in a couple of months then anyways.

Ma27 replied:
Do you still remember when was the last time you've observed it? Perhaps it got fixed in systemd since then (would check the NEWS from it, but I'd like to know where to start approximately).

dali99 replied:
The commit where it was added to the module this was ported from was Oct 10, 2022 so probably then at least.

It could also be funky stuff with nixos-rebuild generating new service files with a pkg-upgrade or something. I recall it being hard to reproduce

Ma27 replied:
> It could also be funky stuff with nixos-rebuild generating new service files with a pkg-upgrade or something

@dali99 If `switch-to-configuration.pl` replaces units (and they have differences) a restart (in some cases a reload, but that's not relevant here) will be done. I'm very sure that this happens within the same transaction. The only exception would be if you have things that get stopped and started later on whereas others get restarted, that happens in multiple transactions IIRC, but this doesn't seem to be the case here.

@999eagle you're running this module for the longest time I guess. Just wondering do you remember any issues with synapse (with workers) getting up again after a deployment?

----

My problem is that right now everything seems to indicate that this isn't a problem anymore and systemd with the config we have in this PR works fine. I guess I'll ask in the nixos systemd channel if someone who probably has more inside knowledge about systemd can take a look and share their opinion, then we'll see how to proceed.

Ma27 replied:
So, while I'm rather certain that the ordering stuff is fine here(Type=notify, `sd_notify(READY=1)` in synapse main happens _after_ successful schema migrations, restart of workers+synapse happens in a single transaction in switch-to-configuration.pl, some of the moving parts here might've changed over time), I don't want to discard the reports entirely.

First of all I've requested in the nixos/systemd channel on Matrix to take a look here, perhaps I've missed something.

After chatting with @mweinelt a bit, we agreed on the following approach:

* We'll observe this for a bit (i.e. await the next synapse bump with schema changes). We have at least two testers (@999eagle and me), though I'm happy for more folks to try this out!
* If we don't encounter any issues with workers failing to start up, we'll merge this into unstable.
* If there's still a case left that'll fail, there's a good chance that we'll catch this before the next branch-off.

My problem is that if we add this now, we'll have a bunch of code which is most likely a no-op (but as I said I may be wrong) that is hard to remove because we _think_ that it might cause issues that are hard to reproduce. That's why I'd like to have this cleared up now.

999eagle replied:
> do you remember any issues with synapse (with workers) getting up again after a deployment?
> under which circumstances did the workers actually crash?

I think I had one crash like this a few months back when a new Synapse version required a database schema upgrade. Workers actually failed to start because the main process hadn't upgraded the database yet.

I haven't observed that behavior in quite a while though (despite multiple database schema upgrades happening), so it's either been fixed in some way by Synapse (probably by moving the `sd_notify` to happen after the schema upgrade is done) or it's a really weird edge case that's very hard to reproduce. I'm hoping it's the former, making a workaround for this no longer required.

mweinelt replied:
I'm also on board with testing, since I've just migrated to this PR.

Migrating from 1.89 to 1.91.2 forced a schema migration on the main process and during the switch-to-configuration all synapse units were stopped and started again. The order of units is not even honored inside a single transaction from what @Ma27 and me could see. While colmena told me the switch had failed, the units however recovered quickly and everything came up as needed.

dali99 replied:
>While colmena told me the switch had failed, the units however recovered quickly

This is indicative of it just restarting the units after failure instead of waiting.

There might be timing related issues where depending on the size of the database, or just the size of the migration The systemd timeout will get reached. And it will stop trying to restart the worker

dali99 replied:
I will turn off my wait-for-synapse script as well to help with the datagathering

Ma27 replied:
> The order of units is not even honored inside a single transaction from what @Ma27 and me could see

It's slightly weird that this didn't happen on my experiments.

But anyways, for reasons I have enough of synapse for today. @mweinelt has a setup where the issue can be observed and there's a new release out with a schema change, so there are plenty of data points now for me to investigate. Will try to do that tomorrow, then we'll see how to proceed.

999eagle replied:
I've updated my Synapse instance to 1.91 and actually ran into this issue again. Workers were started before database migration was done and immediately exited with a failure. Simply starting them again fixed the issue though. The underlying issue with unit dependencies not being handled correctly is most likely an issue with either `systemd` itself or `switch-to-configuration.pl`.
To fix this for `matrix-synapse` I think the cleaner solution would be to add `RestartSec=1s` to the services and maybe set `Restart=always` instead of `Restart=on-failure`. This would allow for easier overriding for users of this module and would cause systemd to handle the wait before automatically restarting the worker services instead of having to handle this manually in a script.

mweinelt replied:
The issue is that, as far as I remember it, systemd only guarantees dependency resolution and ordering during system bring-up, not when you call into `systemctl restart matrix-synapse.service matrix-synapse-worker-foo.service matrix-synapse-worker-bar.service`. They would all be started at the same time. And that is exactly where `switch-to-configuration.pl` comes up short.

999eagle replied:
As far as I can tell, services are actually stopped and started instead of restarted during a switch. So `systemd` is apparently ignoring unit dependencies even during `systemctl start`. Another option to work around this might be to start `matrix-synapse.target` instead of all services individually as `systemd` does respect unit dependencies then.
Can we make `switch-to-configuration.pl` stop all services and the target and then start only the target?

dali99 replied:
In the same [commit](https://github.com/dali99/nixos-matrix-modules/commit/ccbd09267948d5f51ed0fbb1d799a1aa63c4bdbf) as I added the wait-for-script I also added a requires on matrix-synapse to the workers like you did in 24f6a70 so maybe this is indeed enough.

I've been running my module with only the `requires` since last week and it didn't fail with upgrades during that time, for some anecdotal evidence

Ma27 replied:
Just deployed 1.92.1 with @999eagle's latest commit applied and it looks good to me:

```
Sep 18 15:48:39 soost systemd[1]: Starting Synapse Matrix homeserver...
Sep 18 15:48:49 soost systemd[1]: Started Synapse Matrix homeserver.
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker client-0...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker client-1...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-receiver-0...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-receiver-1...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-sender-0...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker federation-sender-1...
Sep 18 15:48:49 soost systemd[1]: Starting Synapse Matrix worker misc...
Sep 18 15:48:55 soost systemd[1]: Started Synapse Matrix worker federation-sender-1.
Sep 18 15:48:55 soost systemd[1]: Started Synapse Matrix worker client-0.
Sep 18 15:48:55 soost systemd[1]: Started Synapse Matrix worker federation-receiver-0.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker federation-sender-0.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker misc.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker federation-receiver-1.
Sep 18 15:48:56 soost systemd[1]: Started Synapse Matrix worker client-1.
```

This time not only via `systemctl restart`, but with `switch-to-configuration.pl` (i..e an actual deploy).

> I've been running my module with only the requires since last week and it didn't fail with upgrades during that time, for some anecdotal evidence

While I'm not 100% convinced by the explanation about how systemd falls short here, I don't think that matters anymore. We have three confirmations that the current behavior is fine now (and what I'd consider desirable).

Even if this falls short, we have a restart behavior in place that will catch that. The only case would be a database that's so slow to migrate that the start timeout (or in now unlikely cases the restart threshold if the dependency handling actually breaks) gets exceeded. However I think that in such a case it's the job of an administrator to increase the timeouts accordingly. Also, I think that restart=always with a configuration that will never get into the StartLimit Burst (i.e. that systemd gives up on start/restart) is not a good idea since this will break monitoring that relies on the failed state of a service (e.g. prometheus-node-exporter alerts) because the failed state will only be reached for a very short period of time.

All in all, I think we can resolve this now.

mweinelt commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -154,7 +197,108 @@ in {
+                  "openid"
+                  "replication"
+                  "static"
+                  "health"
```
Insert in sorted order.

mweinelt reviewed (commented):
This is looking pretty good. We should probably mention that most worker apps require a proper reverse proxy configuration.

999eagle committed:
nixos/synapse: add documentation for required reverse proxy setup

999eagle commented:
Rebased onto `nixpkgs-unstable` to resolve merge conflicts.

999eagle committed:
nixos/synapse: make sure workers require main process

This should ensure systemd handles starting all services (main and
workers) in a single transaction, thus preserving unit orderings
defined through After= even when not restarting the target.

Ma27 commented:
@999eagle thanks!

SuperSandro2000 commented:
Can someone give me a hint how to append the log_config option now? Before I had the following:

```nix
        log_config = pkgs.writeText "synapse-log_config.yaml"
          (lib.readFile (pkgs.path + "/nixos/modules/services/matrix/synapse-log_config.yaml") + ''
            loggers:
              synapse.http.matrixfederationclient:
                level: ERROR

            root:
              level: WARN
              handlers: [journal]
          '');
```
